sdk/log/batch.go

```diff
@@ -0,0 +1,122 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package log // import "go.opentelemetry.io/otel/sdk/log"
+
+import (
+	"context"
+	"time"
+)
+
+// Compile-time check BatchingProcessor implements Processor.
+var _ Processor = (*BatchingProcessor)(nil)
+
+// BatchingProcessor is an processor that asynchronously exports batches of log records.
+type BatchingProcessor struct{}
+
+type batcherConfig struct{}
+
+// NewBatchingProcessor decorates the provided exporter
+// so that the log records are batched before exporting.
+//
+// All of the exporter's methods are called from a single dedicated
+// background goroutine. Therefore, the expoter does not need to
+// be concurrent safe.
+func NewBatchingProcessor(exporter Exporter, opts ...BatchingOption) *BatchingProcessor {
+	// TODO (#5063): Implement.
+	return nil
+}
+
+// OnEmit batches provided log record.
+func (b *BatchingProcessor) OnEmit(ctx context.Context, r Record) error {
+	// TODO (#5063): Implement.
+	return nil
+}
+
+// Shutdown flushes queued log records and shuts down the decorated expoter.
+func (b *BatchingProcessor) Shutdown(ctx context.Context) error {
+	// TODO (#5063): Implement.
+	return nil
+}
+
+// ForceFlush flushes queued log records and flushes the decorated expoter.
+func (b *BatchingProcessor) ForceFlush(ctx context.Context) error {
+	// TODO (#5063): Implement.
+	return nil
+}
+
+// BatchingOption applies a configuration to a Batcher.
+type BatchingOption interface {
+	apply(batcherConfig) batcherConfig
+}
+
+type batchingOptionFunc func(batcherConfig) batcherConfig
+
+func (fn batchingOptionFunc) apply(c batcherConfig) batcherConfig {
+	return fn(c)
+}
+
+// WithMaxQueueSize sets the maximum queue size used by the Batcher.
+// After the size is reached log records are dropped.
+//
+// If the OTEL_BLRP_MAX_QUEUE_SIZE environment variable is set,
+// and this option is not passed, that variable value will be used.
+// If both are set, OTEL_BLRP_MAX_QUEUE_SIZE will take precedence.
```

Comment:
Did we intentionally have different precedence orders for different config?  e.g. https://github.com/open-telemetry/opentelemetry-go/blob/6bc8314a5553ba97e4cfaf7838007085a6be2dea/exporters/otlp/otlpmetric/otlpmetricgrpc/doc.go#L20 seems to have the reverse precedence order

Reply:
(non-blocking comment)

Reply:
TBH I have no idea why I ended up with this. I will update the comments and prototype today.